---
title: "cegwas2 report for TRAIT_NAME_HOLDER"
output: html_document
date: "`r format(Sys.time(), '%d %B, %Y')`"
---


## need to replace my notes with actual report information for users.

1. This R Markdown file should be able to find all data tables and knit out of the box.
<br>
2. It will generate plots (almost) identical to the plots in Analysis folder.
<br>
3. ggsave code is kept for users to make customized plots. 
<br>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
```



```{r include=FALSE}
library(tidyverse)
library(ggbeeswarm)
library(rrBLUP)
library(plotly)
library(DT)
```



```{r echo=FALSE}
# load trait name
# in nextflow use sed to edit this field and make a copy of the .rmd for each trait
trait_name="TRAIT_NAME_HOLDER"   
#trait_name="PC1"

# load independent tests result
total_independent_tests <- read.table("Genotype_Matrix/total_independent_tests.txt", quote="\"", comment.char="", stringsAsFactors=FALSE)

independent_tests <- total_independent_tests[[1]]

independent_test_cutoff <- -log10(0.05/independent_tests)

```


<br>
<br>

## Manhatton plot
* #### Every dot is a SNV marker (??).

* #### Red/pink dots are significant QTL loci based on defined thresholds. Red is BF or user supplied cutoff: default BF -log10(p) > 6. Pink is Eigen cutoff: -log10(0.05/independent_tests) 
<br>
where "independent_tests" is calculated from Eigen Decomposition of SNV correlation matrix, stored in Analysis_Results-date/‚Å©Genotype_Matrix/total_independent_tests.txt

* #### Light blue regions are QTL intervals that contain markers significantly associate with the trait. This is the region that will be shown in the pheno-geno, fine mapping and gene plots below. 


```{r, echo=FALSE, fig.height=3.5, fig.width=10, out.height=350, out.width=1000}
# load processed mapping data. 
# Note that readr::read_delim will throw parsing errors and put NA in rows that contain actual values in strain/value/allele/var.exp so did not use it
processed_mapping <- read.delim(paste0("Mappings/Data/",trait_name,"_processed_mapping.tsv"), stringsAsFactors=FALSE)


# each trait has a separate processed_mapping file now. So the plotting function and loop is removed
# but do check there is only 1 trait and if not, issue warning:
num_traits = length(unique(dplyr::select(processed_mapping,trait)))

if(num_traits > 1){
    print("More than 1 trait in processed_mapping table. Only the first one will be plotted.")
}



# Generate _manplot from Run_Mappings.R

manhattan_plot = processed_mapping %>%
    dplyr::filter(trait == unique(processed_mapping$trait)[1] & CHROM != "MtDNA") %>%
    dplyr::distinct(marker, .keep_all = T) %>%
    dplyr::mutate(EIGEN_CUTOFF = independent_test_cutoff) %>%
    dplyr::mutate(EIGEN_SIG = ifelse(log10p > BF, "1", 
                                     ifelse(log10p > EIGEN_CUTOFF, "2", "0")) )%>%
    ggplot2::ggplot(.) +
    ggplot2::aes(x = POS/1e6, y = log10p) +
    ggplot2::scale_color_manual(values = c("0" = "black", 
                                           "1" = "red",
                                           "2" = "hotpink3")) +
    ggplot2::geom_rect(ggplot2::aes(xmin = startPOS/1e6,    # this is the plot boundary for LD and gene plots
                                    xmax = endPOS/1e6,    # this is the plot boundary for LD and gene plots
                                    ymin = 0, 
                                    ymax = Inf, 
                                    fill = "blue"), 
                       color = "blue",fill = "cyan",linetype = 2, 
                       alpha=.3)+
    ggplot2::geom_hline(ggplot2::aes(yintercept = BF),
                        color = "gray", 
                        alpha = .75,  
                        size = 1) +
    ggplot2::geom_hline(ggplot2::aes(yintercept = EIGEN_CUTOFF),
                        color = "gray", 
                        alpha = .75,  
                        size = 1,
                        linetype = 2) +
    ggplot2::geom_point( ggplot2::aes(color= factor(EIGEN_SIG)) ) +
    ggplot2::facet_grid( . ~ CHROM, scales = "free_x" , space = "free_x") +
    ggplot2::theme_bw() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(size = 16),
                   axis.text.y = ggplot2::element_text(size = 16),
                   axis.title.x = ggplot2::element_text(size = 20, face = "bold", color = "black", vjust = -0.3), 
                   axis.title.y = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   strip.text.x = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   strip.text.y = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   plot.title = ggplot2::element_text(size = 24, face = "bold", vjust = 1), 
                   panel.background = ggplot2::element_rect(color = "black",size = 1.2),
                   legend.position = "none",
                   strip.background = element_blank()) +
    ggplot2::labs(x = "Genomic Position (Mb)",
                  y = expression(-log[10](italic(p))))


# save the plot to file
# ggsave(manhattan_plot, 
#        filename = paste0(trait_name, "_manplot.pdf"),
#        height = 4, 
#        width = 12)

# show plot in html output
manhattan_plot
```



```{r echo=FALSE}
# read in peaks for all traits, quit knitting if there are no significant peaks for this trait
# note that QTL_peaks.tsv is generated even if there is no significant peaks

if (file.info("Mappings/Data/QTL_peaks.tsv")$size!=0) {

#if (file.info("Mappings/Data/QTL_peaks_2row.tsv")$size!=0) {
    
    QTL_peaks <- read.delim("Mappings/Data/QTL_peaks.tsv", header=FALSE, stringsAsFactors=FALSE) %>% 
        dplyr::filter(V1==trait_name)  %>% 
        dplyr::rename(Trait=V1, Chrom=V2, Region_start=V3, Peak=V4, Region_end=V5) %>% 
        dplyr::mutate(Region_width=Region_end - Region_start) %>% 
        dplyr::filter(Region_width < 1500000)
    
    if (nrow(QTL_peaks)==0) {
        
        print("All QTL peaks are too wide for plotting")
        
        whether_fine_map = FALSE
        
    } else {
        
        # define size of plot based on number of peaks
        plot_height_all_peaks = 350*nrow(QTL_peaks) 
        
        # set parameter to run all fine-mapping related chuncks
        whether_fine_map = TRUE
    }    
    
    
} else { 
    print("No significant peak for this trait")
    whether_fine_map = FALSE
}
```



```{r eval=whether_fine_map}
# print out the table to html
# this doesn't print within if{}
DT::datatable(QTL_peaks)
```


<br>
<br>

## Interactive pxgplot, fine mapping plot, gene plot for each peak region
* #### Select an area to zoom in, double click anywhere on the plot to reset zoom.
* #### the snpeff table is too long. group by marker???
```{r echo=FALSE, out.height=500, out.width=1000, eval=whether_fine_map}
# first create a list to put plotly plots and DT tables since they cannot be directly printed within for loops in r markdown
# plotly_pxg_fine_list <- htmltools::tagList()
plotly_pxg_fine_list <- list()
DT_gene_list <- list()
plotly_gene_plot_list <- list()

# read in data table for gene annotation for each QTL region
tidy_genes_in_region <- read.delim(paste0("Fine_Mappings/Data/", trait_name, "_snpeff_genes.tsv"), stringsAsFactors=FALSE) %>% 
    select(MARKER, POS, REF, ALT, GENE_NAME, WBGeneID, VARIANT_IMPACT, PEAK_MARKER, CHROM, STRAND, TRANSCRIPTION_START_POS, TRANSCRIPTION_END_POS, QTL_INTERVAL_START, QTL_INTERVAL_END, VARIANT_LOG10p, STRAIN, STRAIN_GENOTYPE, Phenotype_Value)
# [1] "MARKER"                      "CHROM"                       "POS"                        
#  [4] "REF"                         "ALT"                         "MAF_variant"                
#  [7] "GENE_NAME"                   "WBGeneID"                    "WBFeature_TYPE"             
# [10] "WBFeature_ID"                "TRANSCRIPT_BIOTYPE"          "VARIANT_IMPACT"             
# [13] "NUCLEOTIDE_CHANGE"           "AMINO_ACID_CHANGE"           "STRAND"                     
# [16] "TRANSCRIPTION_START_POS"     "TRANSCRIPTION_END_POS"       "PEAK_MARKER"                
# [19] "PEAK_MAF"                    "TRAIT"                       "QTL_INTERVAL_START"         
# [22] "QTL_INTERVAL_END"            "VARIANT_LD_WITH_PEAK_MARKER" "VARIANT_LOG10p"             
# [25] "STRAIN"                      "STRAIN_GENOTYPE"             "Phenotype_Value"   

if (length(unique(tidy_genes_in_region$QTL_INTERVAL_START)) != nrow(QTL_peaks)) {
    print("number of significant QTL regions don't match!!!")
}

# For each peak, make 1 row of plots  
for (n in seq_len(nrow(QTL_peaks))) {
    
    peak_name = paste0(QTL_peaks[n, "Trait"],".",QTL_peaks[n, "Chrom"],".",QTL_peaks[n, "Region_start"],".",QTL_peaks[n, "Region_end"])
    
    ############## phenotype-genotype plot from Run_Mapping.R
    pxg_df <- na.omit(processed_mapping) %>% # probably the na.omit is not needed any more
        dplyr::filter(CHROM==QTL_peaks[n, "Chrom"], peakPOS==QTL_peaks[n, "Peak"]) %>%  
        dplyr::mutate(value=as.numeric(value)) %>% 
        group_by(allele) %>%
        dplyr::mutate(mean_pheno = mean(value, na.rm = T)) 
    
    
    pxg_plot <- pxg_df %>% 
        ggplot()+
        aes(x = factor(allele, levels = c(-1,1), labels = c("REF","ALT")))+
        geom_beeswarm(cex=4, priority='density',
                      aes(y = value, plotly_label=strain),
                      shape = 21, 
                      fill = "gray50",
                      size = 2)+
        geom_point(aes(y = mean_pheno), 
                   fill = "red", 
                   size = 3, 
                   shape = 25)+
        theme_bw(15)+
        labs(y = trait_name,
             x = "Genotype") +
        theme(axis.text.x = ggplot2::element_text(size = 14),
              axis.text.y = ggplot2::element_text(size = 14),
              axis.title.x = ggplot2::element_text(size = 16, face = "bold", color = "black", vjust = -0.3),
              axis.title.y = ggplot2::element_text(size = 16, face = "bold", color = "black", vjust = -0.3),
              strip.text.x = element_text(size = 14, face = "bold"))+
        theme(legend.position = "none") +
        ggtitle(peak_name)
    
    
    # save plot to file
    # ggsave(pxg_plot, filename = paste0(peak_name, "_pxgplot.pdf"),
    #        height = 4, 
    #        width = 4, limitsize = FALSE)
    
    
    # store plot for html output
    plotly_pxg_fine_list[[1+(n-1)*2]] <- ggplotly(pxg_plot, tooltip = c("plotly_label", "y"))
    
    
    
    #################### Generate fine mapping plots from Finemap_QTL_Intervals.R
    
    # load data. File name format is PC1.II.7319019.8506109_prLD_df.tsv
    prLD_file_name = paste0("Fine_Mappings/Data/",QTL_peaks[n, "Trait"],".",QTL_peaks[n, "Chrom"],".",QTL_peaks[n, "Region_start"],".",QTL_peaks[n, "Region_end"],"_prLD_df.tsv")
    
    pr_roi_ld <- read.delim(prLD_file_name, stringsAsFactors=FALSE) %>% 
        dplyr::mutate(marker = stringr::str_replace(marker, "_", ":")) 
    
    peak_roi_marker <- dplyr::filter(pr_roi_ld, POS == QTL_peaks[n, "Peak"])
    
    ld_plot <- pr_roi_ld %>% ggplot() +
        aes(x = POS/1e6, plotly_label_ld = marker) +
        geom_point(aes(color = ld_r2, y = value), shape = 17, size = 2) +
        geom_point(data = peak_roi_marker, aes(y = value), shape = 23, size = 2, fill = "red") +
        scale_color_viridis_c(name = "R2") +
        theme_bw(15)+
        labs(x = "Genomic Position (Mb)",
             y = "-log[10]p")
    # need to write out a separate plot with stroke around diamond points
    
    
    # save plot to file
    # ggsave(ld_plot, filename = paste0(peak_name, "_finemap_plot.pdf"),
    #        height = 4,
    #        width = 12)
    # 
    
    # store plot for html output
    # cannot get plotly to work with fill with continuous scale, so use a shape with no stroke(color)/fill distinction.
    plotly_pxg_fine_list[[2+(n-1)*2]] <- hide_colorbar(ggplotly(ld_plot, tooltip = c("plotly_label_ld", "y"))) 
    
    
    
    
    ######################## Generate gene plot from plot_genes.R
    # create table to store gene related information
    gene_info <- tidy_genes_in_region %>%
        dplyr::filter(QTL_INTERVAL_START == QTL_peaks$Region_start[n] & !is.na(WBGeneID)) %>%
        dplyr::select(GENE_NAME, WBGeneID, CHROM, STRAND, TRANSCRIPTION_START_POS, TRANSCRIPTION_END_POS, QTL_INTERVAL_START, QTL_INTERVAL_END) %>% dplyr::distinct()
    
    # for each gene, keep the most significant overlapping variant log10p
    # this is used as y-axix of gene position
    gene_df <- tidy_genes_in_region %>%
        dplyr::filter(QTL_INTERVAL_START == QTL_peaks$Region_start[n] & !is.na(WBGeneID)) %>%
        dplyr::distinct(MARKER, GENE_NAME, WBGeneID, PEAK_MARKER, CHROM, STRAND, TRANSCRIPTION_START_POS, TRANSCRIPTION_END_POS, QTL_INTERVAL_START, QTL_INTERVAL_END, VARIANT_LOG10p) %>%
        dplyr::group_by(WBGeneID) %>% 
        # the 2 lines below are used in plot_genes.R instead
        # dplyr::mutate(VARIANT_LOG10p = max(VARIANT_LOG10p, na.rm = T)) %>%
        # dplyr::distinct()
        dplyr::summarise(VARIANT_LOG10p = max(VARIANT_LOG10p, na.rm = T)) %>% 
        dplyr::inner_join(gene_info) %>% 
        dplyr::mutate(STRAND = ifelse(STRAND=="+", "Gene_on_plus_strand", "Gene_on_minus_strand"))
    
    
    # peak_variant <- as.numeric(strsplit(unique(gene_df$PEAK_MARKER), split = ":")[[1]][2])
    peak_variant <- QTL_peaks[n, "Peak"]
    
    variant_df <- tidy_genes_in_region %>%
        dplyr::filter(QTL_INTERVAL_START == QTL_peaks[n, "Region_start"]) %>%
        dplyr::distinct(CHROM, POS, VARIANT_LOG10p, VARIANT_IMPACT, REF, ALT) %>% 
        dplyr::mutate(MARKER_pos = paste0(CHROM, POS, " REF:ALT ", REF, ":", ALT))
    
    variant_df$VARIANT_IMPACT[is.na(variant_df$VARIANT_IMPACT)] <- "Intergenic"
    
    xs <- unique(gene_df$QTL_INTERVAL_START)
    xe <- unique(gene_df$QTL_INTERVAL_END)
    gc <- unique(gene_df$CHROM)
    
    max_logp <- unique(max(variant_df$VARIANT_LOG10p))/150
    
    gene_plot <- ggplot(gene_df) +
        geom_vline(aes(xintercept = peak_variant/1e6),
                   linetype=3, color = "cyan")+
        geom_segment(aes(x = ifelse(STRAND == "Gene_on_plus_strand", TRANSCRIPTION_START_POS/1e6, TRANSCRIPTION_END_POS/1e6),
                         xend = ifelse(STRAND == "Gene_on_plus_strand", TRANSCRIPTION_END_POS/1e6, TRANSCRIPTION_START_POS/1e6),
                         y = VARIANT_LOG10p,
                         yend = VARIANT_LOG10p,
                         color = STRAND,
                         plotly_label_gene_name = GENE_NAME,
                         plotly_label_gene_ID = WBGeneID),
                     size = 1.5) +
        geom_segment(data = variant_df,
                     aes(x = POS/1e6,
                         xend = POS/1e6,
                         y = VARIANT_LOG10p+max_logp,
                         yend = VARIANT_LOG10p-max_logp,
                         color = VARIANT_IMPACT, 
                         plotly_label_variant = MARKER_pos)) +
        scale_color_manual(values = c("Gene_on_plus_strand" = "#3bb5c4",
                                      "Gene_on_minus_strand" = "#341d8d",
                                      "MODIFIER" = "gray50",
                                      "LOW" = "gray30",
                                      "MODERATE" = "orange",
                                      "HIGH" = "red",
                                      "Intergenic" = "gray80"),
                           # labels = c("Gene_on_plus_strand" = "gene on + strand", 
                           #            "Gene_on_minus_strand" = "gene on - strand", 
                           #            "HIGH" = "Variant impact high", 
                           #            "MODERATE" = "Variant impact moderate", 
                           #            "LOW" = "Variant impact low", 
                           #            "MODIFIER" = "Variant modifies protein", 
                           #            "Intergenic" = "Intergenic variant"),
                           name = "Variant Impact")+
        labs(x = "Genomic Position (Mb)",
             y = "-log[10]p") +
        theme_bw(18)+
        xlim(c(xs/1e6, xe/1e6)) +
        theme(legend.position = "top",
              panel.grid = element_blank())
    

    # save plot to file
    # ggsave(gene_plot,
    #       filename = paste0(trait_name, ".", gc, ".",  xs, ".", xe, "_gene_plot.pdf"),
    #        height=10, width = 14)
    
    
    # store plot for html output
    plotly_gene_plot_list[[n]] <- ggplotly(gene_plot, tooltip = c("plotly_label_gene_name", "plotly_label_gene_ID", "plotly_label_variant")) 
    
    
    # store table for html output
    gene_list_out = select(tidy_genes_in_region, MARKER, GENE_NAME, WBGeneID, VARIANT_IMPACT, PEAK_MARKER, VARIANT_LOG10p, STRAIN, STRAIN_GENOTYPE, Phenotype_Value) # %>% dplyr::slice(1:20)
    DT_gene_list[[n]] <- DT::datatable(gene_list_out)
}


```


Need to plot the plotly_plot in a row, then the gene plot, then the table. all in htmlwidgets probably. should do all in 1 loop. 

<br>
<br>


```{r include=FALSE, results='asis'}
# in order to make htmlwidgets print in a for loop in rmarkdown:
# https://github.com/ropensci/plotly/issues/273
# https://github.com/rstudio/DT/issues/67
# initiation is needed to attach packages. put in a separate chunk so the output will not be included in the html
plot_ly()
DT::datatable(matrix())
```


```{r, results='asis', eval=whether_fine_map}
for (i in seq_len(nrow(QTL_peaks))) {
    cat("\n\n")
    print(htmltools::tagList(plotly::subplot(plotly_pxg_fine_list[[2*i-1]], plotly_pxg_fine_list[[2*i]], shareY = FALSE, titleX = TRUE, titleY = TRUE, nrows = 1, margin=0.05, widths=c(0.3, 0.7)))) 
    cat("\n\n")
    
    cat("\nClick on legend to hide/show certain groups\n")
    print(htmltools::tagList(plotly_gene_plot_list[[i]]))
    cat("\n\n")
    
    print(htmltools::tagList(DT_gene_list[[i]]))
    cat("\n\n")
}
```
<br>
<br>

## Burden mapping plots

* #### Every dot represents a gene (???)
* #### SKAT
```{r, echo=FALSE, fig.height=3.5, fig.width=10, out.height=350, out.width=1000}
# plots from plot_burden.R
# Plot SKAT
skat <- read.delim(paste0("BURDEN/SKAT/Data/", trait_name, ".Skat.assoc"), stringsAsFactors=FALSE) %>%
    dplyr::rowwise() %>%
    dplyr::mutate(CHROM = strsplit(strsplit(RANGE,split = ",")[[1]][1],split = ":")[[1]][1],
                  POS = as.numeric(strsplit(strsplit(strsplit(RANGE,split = ",")[[1]][1],split = ":")[[1]][2],split = "-")[[1]][1]),
                  endPOS = as.numeric(strsplit(strsplit(strsplit(RANGE,split = ",")[[1]][1],split = ":")[[1]][2],split = "-")[[1]][2]))%>%
    dplyr::mutate(size = abs(POS-endPOS)) %>%
    dplyr::filter(CHROM!="MtDNA")%>%
    dplyr::ungroup()%>%
    dplyr::filter(NumVar > 1)%>%
    dplyr::mutate(significant = ifelse(Pvalue < .05/n(), TRUE,FALSE ))%>%
    dplyr::arrange(Pvalue)

skat_plot <- skat%>%
    ggplot()+
    aes(x = POS/1e6, y = -log10(Pvalue), alpha = 0.5, color = significant, plotly_label_geneID = Gene )+
    geom_point()+
    scale_color_manual(values=c("black","red"))+
    facet_grid(.~CHROM, scales = "free", space = "free")+
    ggplot2::theme_bw() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(size = 16),
                   axis.text.y = ggplot2::element_text(size = 16),
                   axis.title.x = ggplot2::element_text(size = 20, face = "bold", color = "black", vjust = -0.3), 
                   axis.title.y = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   strip.text.x = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   strip.text.y = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   plot.title = ggplot2::element_text(size = 24, face = "bold", vjust = 1), 
                   panel.background = ggplot2::element_rect(color = "black",size = 1.2),
                   legend.position = "none",
                   strip.background = element_blank())+
    labs(x = "Genomic Position (Mb)", 
         y = "-log[10](p)")

# save plot to file
# ggsave(filename = paste0(trait_name, "_SKAT.pdf"), 
#        plot = skat_plot,
#        height = 4, 
#        width = 12)

# print plot to html output
# print(skat_plot)
ggplotly(skat_plot, tooltip = c("plotly_label_geneID","y")) %>% toWebGL()


# show table in html output. 
# Pvalue is the y-axis in plot
skat_link = filter(skat, significant == TRUE) %>% 
    mutate(Gene_ID = paste0('<a href=https://wormbase.org/species/c_elegans/gene/', Gene, '>', Gene,'</a>' ) ) %>%
    select(Gene_ID, RANGE, Stat, significant)  

DT::datatable(skat_link, escape = FALSE)
```


* #### Variable Threshold
```{r, echo=FALSE, fig.height=3.5, fig.width=10, out.height=350, out.width=1000}
##############
# PLOT VTprice
benreg_burden <- read.delim(paste0("BURDEN/VT/Data/", trait_name, ".VariableThresholdPrice.assoc"), stringsAsFactors=FALSE)
benreg_burden$Gene <- as.character(benreg_burden$Gene)
benreg_burden$RANGE <- as.character(benreg_burden$RANGE)

benreg_burden<-benreg_burden%>%
    dplyr::rowwise()%>%
    dplyr::mutate(CHROM = strsplit(strsplit(RANGE,split = ",")[[1]][1],split = ":")[[1]][1],
                  POS = as.numeric(strsplit(strsplit(strsplit(RANGE,split = ",")[[1]][1],split = ":")[[1]][2],split = "-")[[1]][1]),
                  endPOS = as.numeric(strsplit(strsplit(strsplit(RANGE,split = ",")[[1]][1],split = ":")[[1]][2],split = "-")[[1]][2]))%>%
    dplyr::mutate(size = abs(POS-endPOS)) %>%
    ungroup()%>%
    dplyr::mutate(significant = ifelse(PermPvalue < .05/n(), TRUE,FALSE ))%>%
    dplyr::filter(CHROM!="MtDNA",NumVar>1,size >500)

burden_manplot <- benreg_burden%>%
    ggplot()+
    aes(x = POS/1e6, y = Stat, size = NumVar, alpha = 0.5, color = significant, plotly_label_geneID = Gene) +
    geom_point(size = 0.25)+
    scale_color_manual(values=c("black","red"))+
    facet_grid(.~CHROM, scales = "free", space = "free")+
    theme_bw()+
    ggplot2::theme_bw() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(size = 16),
                   axis.text.y = ggplot2::element_text(size = 16),
                   axis.title.x = ggplot2::element_text(size = 20, face = "bold", color = "black", vjust = -0.3), 
                   axis.title.y = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   strip.text.x = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   strip.text.y = ggplot2::element_text(size = 20, face = "bold", color = "black"), 
                   plot.title = ggplot2::element_text(size = 24, face = "bold", vjust = 1), 
                   panel.background = ggplot2::element_rect(color = "black",size = 1.2),
                   legend.position = "none",
                   strip.background = element_blank())+
    labs(x = "Genomic Position (Mb)", y = "Test Statisitic")

# save plot to file
# ggsave(filename = paste0(trait_name, "_VTprice.pdf"), 
#        plot = burden_manplot,
#        height = 4, 
#        width = 12)

# show plot in html output
# print(burden_manplot)
ggplotly(burden_manplot, tooltip = c("plotly_label_geneID","y")) %>% toWebGL()


# show table in html output. 
# Stat is the y-axis in plot. what is it??
benreg_burden_link = filter(benreg_burden, significant == TRUE) %>% 
    mutate(Gene_ID = paste0('<a href=https://wormbase.org/species/c_elegans/gene/', Gene, '>', Gene,'</a>' ) ) %>%
    select(Gene_ID, RANGE, Stat, significant)  

DT::datatable(benreg_burden_link, escape = FALSE)
```


* #### Genes significant in both statistical tests 
```{r}
skat_sig = dplyr::mutate(skat_link, SKAT_stat = Stat)

benreg_burden_sig = dplyr::mutate(benreg_burden_link, VariableThreshold_stat = Stat)

both_sig = inner_join(skat_sig, benreg_burden_sig, by=c("Gene_ID","RANGE"))

DT::datatable(both_sig) 

```


```{r}
sessionInfo()
```
