---
title: "cegwas2 report for TRAIT_NAME_HOLDER: QTL_CHROM_HOLDER.QTL_REGION_START_HOLDER-QTL_REGION_END_HOLDER"

date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: true
      smooth_scroll: true
---

<br>

## Usage

* This R Markdown file imports data tables within Analysis_Results-Date folder and will knit without any modification.
<br>
* Features for the interactive plots: 
* Hover over the plot region to show underlying data.
* Click and drag to select a zoom region, double click to reset zoom.
* Click on the legend to show/hide different data groups.
<br>
* ggsave code is kept for users to make customized plots. 
<br>
<br>
<br>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
```



```{r include=FALSE}
library.path <- .libPaths()
library("tidyverse", lib.loc = library.path)
library("ggbeeswarm", lib.loc = library.path)
library("rrBLUP", lib.loc = library.path)
library("plotly", lib.loc = library.path)
library("cowplot", lib.loc = library.path)
library("ComplexHeatmap", lib.loc = library.path)

```


```{r echo=FALSE}
# set trait name
# in nextflow use sed to edit this field and make a copy of the .rmd for each trait
trait_name <- "TRAIT_NAME_HOLDER"

# set QTL region in QTL_peaks.tsv
# in nextflow use sed to edit this field and make a copy of the .rmd for each trait and each QTL region
QTL_Chrom <- "QTL_CHROM_HOLDER"
QTL_Region_start <- QTL_REGION_START_HOLDER
QTL_Peak <- QTL_PEAK_HOLDER
QTL_Region_end <- QTL_REGION_END_HOLDER

```


```{r echo=FALSE}
# load independent tests result
total_independent_tests <- read.table("Genotype_Matrix/total_independent_tests.txt", quote="\"", comment.char="", stringsAsFactors=FALSE)

independent_test_cutoff <- -log10(0.05/total_independent_tests[[1]])

# load processed mapping data. 
# Note that readr::read_delim will throw parsing errors and put NA in rows that contain actual values in strain/value/allele/var.exp so did not use it
processed_mapping <- read.delim(paste0("Mappings/Data/",trait_name,"_processed_mapping.tsv"), stringsAsFactors=FALSE)

user_strain_pheno_list <- processed_mapping %>%
    dplyr::select(strain, value) %>%
    dplyr::distinct() %>% 
    na.omit()
```



<br>
<br>

## Phenotype x Genotype plot:
```{r echo=FALSE}
# allow for bcsq or snpeff
file <- grep("genes.tsv", grep(glue::glue("{trait_name}"), list.files("~/Downloads"), value = T), value = T)

tidy_genes_in_region <- read.delim(paste0("~/Downloads/", file), stringsAsFactors=FALSE) %>% 
    dplyr::select(MARKER, POS, REF, ALT, GENE_NAME = GENE, WBGeneID, VARIANT_IMPACT, PEAK_MARKER, CHROM, STRAND, TRANSCRIPTION_START_POS, 
                  TRANSCRIPTION_END_POS, QTL_INTERVAL_START, QTL_INTERVAL_END, VARIANT_LOG10p, STRAIN, STRAIN_GENOTYPE)

QTL_peak <- paste0(QTL_Chrom,":",QTL_Peak)

QTL_region <- paste0(QTL_Chrom,":",QTL_Region_start,"-",QTL_Region_end)

############## phenotype-genotype plot from Run_Mapping.R
pxg_df <- processed_mapping %>%
    na.omit() %>% # probably the na.omit is not needed any more
    dplyr::filter(CHROM == QTL_Chrom, peakPOS == QTL_Peak) %>%  
    dplyr::mutate(value = as.numeric(value)) %>% 
    group_by(allele) %>%
    dplyr::mutate(mean_pheno = mean(value, na.rm = T)) 


pxg_plot <- pxg_df %>% 
    ggplot()+
    aes(x = factor(allele, levels = c(-1,1), labels = c("REF","ALT")))+
    geom_beeswarm(cex=3, priority='density',
                  aes(y = value, plotly_label=strain),
                  shape = 21, 
                  fill = "gray50",
                  size = 2)+
    geom_point(aes(y = mean_pheno), 
               fill = "red", 
               size = 3, 
               shape = 25)+
    theme_bw(15)+
    labs(y = trait_name,
         x = "Genotype") +
    theme(axis.text.x = ggplot2::element_text(size = 14),
          axis.text.y = ggplot2::element_text(size = 14),
          axis.title.x = ggplot2::element_text(size = 16, face = "bold", color = "black", vjust = -0.3),
          axis.title.y = ggplot2::element_text(size = 16, face = "bold", color = "black", vjust = -0.3),
          strip.text.x = element_text(size = 14, face = "bold"))+
    theme(legend.position = "none") 


# save plot to file
# ggsave(pxg_plot, filename = paste0(QTL_peak, "_pxgplot.pdf"),
#        height = 4, 
#        width = 4, limitsize = FALSE)


# interactive plots
plotly::ggplotly(pxg_plot, tooltip = c("plotly_label", "y")) %>%
    layout(autosize = F, width = 400, height = 400)

```

<br>
<br>

## Divergent region


```{r}
# check whether there are user tested strains that have no divergent region data
div_strain_list <- read.delim("Divergent_and_haplotype/div_isotype_list.txt", header=FALSE, stringsAsFactors=FALSE) 

w_div_data_strain <- user_strain_pheno_list %>%
    dplyr::inner_join(div_strain_list, by=c("strain"="V1"))
no_data_strain <- user_strain_pheno_list %>%
    dplyr::anti_join(div_strain_list, by=c("strain"="V1"))

print("User-tested strains that do not have divergent region data (shown in blue below): ")
print(no_data_strain$strain)
```


```{r}
# genome bin table
all_QTL_bins <- read.delim("Divergent_and_haplotype/all_QTL_bins.bed", header=FALSE, stringsAsFactors=FALSE)
```




```{r}
# take genomic 1kb bins within the QTL region
# format to the same as div regions
# during spread, this "all_bins" will dictate number of rows in resulting dataframe
region_QTL_bins <- all_QTL_bins %>%
    dplyr::filter(V1==QTL_Chrom & V2 > QTL_Region_start & V3 < QTL_Region_end ) %>% 
    dplyr::mutate(V4="all_bins", V5=0)


# if any of user supplied strains have divergent region with the QTL region
if (file.info("Divergent_and_haplotype/all_QTL_div.bed")$size!=0) {
    
    all_QTL_div <- read.delim("Divergent_and_haplotype/all_QTL_div.bed", header=FALSE, stringsAsFactors=FALSE)
    
} else {
    # create an empty dataframe
    all_QTL_div <- data.frame(V1=character(),
                              V2=integer(), 
                              V3=integer(), 
                              V4=character(), 
                              V5=integer(), 
                              stringsAsFactors=FALSE)
}


strain_QTL_div <- all_QTL_div %>%
    dplyr::filter(V4 %in% user_strain_pheno_list$strain)


# take divergent bins within the QTL region, for all user tested strains
# append all_bins within QTL region to the QTL_div 
# spread to short form so each strain is a column, each bin in QTL region is a row, all bins are included.
region_QTL_div <- strain_QTL_div %>%
    dplyr::filter(V1==QTL_Chrom & V2 > QTL_Region_start & V3 < QTL_Region_end ) %>%
    dplyr::bind_rows(region_QTL_bins)  %>% 
    tidyr::spread(V4, V5, fill=0) %>% 
    dplyr::select(-all_bins)

# add back user-tested strains that have divergent data but are not divergent within the QTL peak region, therefore left out in previous step
w_div_data_strain_missed <- w_div_data_strain$strain[!(w_div_data_strain$strain %in% names(region_QTL_div))]


for (x in w_div_data_strain_missed){
    region_QTL_div[[x]] <- 0
}


# add back user-tested strains that have no divergent data
for (x in no_data_strain$strain){
    region_QTL_div[[x]] <- -1
}


# check that all user-tested strains are include now
if (ncol(region_QTL_div) - 3 != nrow(user_strain_pheno_list)){ print("strain number don't match!!!") }


m <- as.matrix(dplyr::select(region_QTL_div,-V1,-V2,-V3))

# transpose to the format that heatmap needs
m2 <- t(m)
colnames(m2) <- region_QTL_div$V2


# add REF/ALT annotation
ref_alt <- processed_mapping %>%
    dplyr::filter(peakPOS == QTL_Peak) %>% 
    dplyr::select(strain, allele) %>% 
    dplyr::filter(strain %in% row.names(m2))

alt_count <- data.frame(table(ref_alt$allele)) %>% 
    dplyr::filter(Var1==1)
ref_count <- data.frame(table(ref_alt$allele)) %>% 
    dplyr::filter(Var1==-1)

ref_alt <- ref_alt %>%
    dplyr::mutate(geno_group = ifelse(allele==1, "ALT", "REF"))  %>% 
    dplyr::mutate(geno_group_count = ifelse(allele==1, paste0("ALT(", alt_count[1,2], ")"), paste0("REF(", ref_count[1,2], ")"))) %>% 
    dplyr::select(-allele)

# make sure REF/ALT annotation is in the same order as row.names(m2)
# that's how complexHeatmap combine them
ref_alt_annotation <- data.frame(strain=row.names(m2)) %>% 
    dplyr::left_join(ref_alt, by=c("strain" = "strain"))


# add div and write out table
# this corresponds to the sort order of rows for heatmap
peak_bin <- floor(QTL_Peak/1000)*1000
ref_alt_annotation <- ref_alt_annotation %>%
    dplyr::rowwise() %>%
    mutate(is_div = m2[strain, as.character(peak_bin)])


# decide whether there is enough strains with divergent region to split phenotype genotype plot
if (nrow(dplyr::filter(ref_alt_annotation, is_div == 1)) > nrow(ref_alt_annotation)/20 ) {
    split_by_div = TRUE
} else { split_by_div = FALSE }

######################### make the plot
# set color for heatmap
colors <- structure(c("indianred1","grey90","dodgerblue3"), names = c("1", "0", "-1")) 

# keep only existing groups for legend
legend_mapping <- data.frame(div_group=c(0, 1, -1), div_label=c("not divergent", "divergent from N2", "no divergent data" )) %>% 
    dplyr::filter(div_group %in% unique(c(m2)))

# create x-axis label
x_axis_start <- colnames(m2)[1]
x_axis_peak <- round(QTL_Peak, digits = -3)
x_axis_end <- colnames(m2)[ncol(m2)]

genomic_coor_text <- as.numeric(colnames(m2))
genomic_coor_text[!genomic_coor_text %in% c(x_axis_start,x_axis_peak, x_axis_end)] <- ""

x_axis_label <- ComplexHeatmap::HeatmapAnnotation(
    year = anno_text(genomic_coor_text, rot = 0, location = unit(1, "npc"), just = "top"))

# add empty space on top of plot so the text "peak" will be within plottin area and show
peak_label_space <- ComplexHeatmap::HeatmapAnnotation(foo = anno_empty(border = FALSE, height = unit(1, "cm")))


ht <- ComplexHeatmap::Heatmap(m2, name="div_region",
              row_split=factor(ref_alt_annotation$geno_group_count,
                               levels=c(paste0("REF(", ref_count[1,2], ")"), paste0("ALT(", alt_count[1,2], ")"))), row_gap = unit(3, "mm"),
              cluster_columns = FALSE, 
              show_row_dend = FALSE, 
              cluster_row_slices=FALSE, 
              col = colors, 
              heatmap_legend_param = list( title="",
                                           at = legend_mapping$div_group,
                                           labels = as.character(legend_mapping$div_label)),
              show_column_names = FALSE,
              show_row_names = FALSE,
              bottom_annotation = x_axis_label,
              top_annotation = peak_label_space)

draw(ht)

ComplexHeatmap::decorate_heatmap_body("div_region", {
    i = which(colnames(m2) == round(QTL_Peak, digits = -3))
    x = i/ncol(m2)
    grid.lines(c(x, x), c(0, 1), gp = gpar(lwd = 2, lty = 2))
    grid.text("QTL peak", x, unit(1, "npc") + unit(5, "mm"))
}, row_slice=1)

ComplexHeatmap::decorate_heatmap_body("div_region", {
    i = which(colnames(m2) == round(QTL_Peak, digits = -3))
    x = i/ncol(m2)
    grid.lines(c(x, x), c(0, 1), gp = gpar(lwd = 2, lty = 2))
}, row_slice=2)

#dev.copy(pdf,'divergent_regions.pdf')
#dev.off()


```

<br>
<br>

## Haplotype
### Two different color schemes were used to plot the same data, to better differentiate different haplotypes.
<br>
```{r, fig.width=20, fig.height=10}
# if (file.info("all_QTL_div.bed")$size!=0) {

haplotype_in_QTL_region <- read.delim("Divergent_and_haplotype/haplotype_in_QTL_region.txt", header=FALSE, stringsAsFactors=FALSE) %>% 
    dplyr::select(-V13)

# names(haplotype_in_QTL_region) <- c("chromosome", "start", "stop", "haplotype", "isotype", "plotpoint", "segment", "color", "color_new", "Chrom", "Region_start", "Region_end")

names(haplotype_in_QTL_region) <- c("chromosome", "start", "stop", "haplotype", "isotype", "plotpoint", "segment", "color", "Chrom", "Region_start", "Region_end")


########### filter only tested strains

########### keep rows for this QTL region
############ keep relavent columns for plotting

plot_df <- haplotype_in_QTL_region %>%
    dplyr::filter(Chrom == QTL_Chrom & Region_start == QTL_Region_start) %>% 
    dplyr::select(chromosome, start, stop, haplotype, isotype, plotpoint, segment, color)


########### sort by haplotype of 1st and last segment

plot_df_seg_1 <- plot_df %>% 
    dplyr::arrange(isotype, start) %>%
    dplyr::distinct(isotype, .keep_all = TRUE) %>% 
    dplyr::select(isotype, haplotype) %>% 
    dplyr::rename(haplotype_start = haplotype)

plot_df_seg_last <- plot_df %>% 
    dplyr::arrange(isotype, -start) %>%
    dplyr::distinct(isotype, .keep_all = TRUE) %>% 
    dplyr::select(isotype, haplotype) %>% 
    dplyr::rename(haplotype_end = haplotype) 

plot_df_sort_order <- plot_df_seg_1 %>%
    dplyr::inner_join(plot_df_seg_last) %>%
    dplyr::arrange(haplotype_start, haplotype_end) %>% 
    dplyr::mutate(plotpoint2=row_number()) %>% 
    dplyr::inner_join(ref_alt, by=c("isotype"="strain")) %>% 
    dplyr::group_by(geno_group) %>% 
    dplyr::mutate(plotpoint3 = rank(plotpoint2)) %>% 
    dplyr::ungroup() %>% 
    dplyr::select(isotype, plotpoint3, geno_group)


########## add new plotpoint back to plot_df 

plot_df <- plot_df %>%
    dplyr::inner_join(plot_df_sort_order, by="isotype")

########## relevel REF/ALT

plot_df$geno_group <- factor(plot_df$geno_group, levels=c("REF","ALT"))


#=======================#
# Normal haplotype plot #
#=======================#


mcolor_grp <- plot_df %>% 
    dplyr::select(haplotype, color) %>% 
    dplyr::distinct()
mcolor <- mcolor_grp$color

# mcolor_grp2 <- plot_df %>% 
#     dplyr::select(haplotype, color_new) %>% 
#     dplyr::distinct()
# mcolor2 <- mcolor_grp2$color_new

names(mcolor) <- mcolor_grp$haplotype

strain_labels <- plot_df %>%
    dplyr::select(isotype, plotpoint3) %>% 
    dplyr::distinct()

plot_df <- plot_df %>%
    dplyr::arrange((plotpoint3))

plot_df$isotype <- factor(plot_df$isotype, unique(plot_df$isotype))

p1 <- ggplot2::ggplot(filter(plot_df),
            ggplot2::aes(x = start, xend = stop,
                 y = isotype, yend = isotype,
                 color = haplotype)) +
    ggplot2::geom_segment(size = 3) +
    ggplot2::scale_color_manual(values = mcolor) +
    ggplot2::xlab("Position") +
    ggplot2::theme_bw() +
    ggplot2::coord_cartesian(xlim=c(QTL_Region_start, QTL_Region_end)) +
    ggplot2::theme(legend.position="none") +
    ggplot2::geom_vline(xintercept = QTL_Peak) +
    ggplot2::facet_grid(geno_group ~ ., scales = "free", space = "free") + 
    ggplot2::theme(panel.spacing = unit(1, "lines"))


#ggsave(paste0("haplotype_", QTL_Chrom, ".", QTL_Region_start, ".", QTL_Region_end, ".png"), p, height = 30, width = 8)


# p2 <- ggplot2::ggplot(filter(plot_df),
#              ggplot2::aes(x = start, xend = stop,
#                  y = isotype, yend = isotype,
#                  color = haplotype)) +
#     ggplot2::geom_segment(size = 3) +
#     ggplot2::scale_color_manual(values = mcolor2) +
#     ggplot2::xlab("Position") +
#     ggplot2::theme_bw() +
#     ggplot2::coord_cartesian(xlim=c(QTL_Region_start, QTL_Region_end)) +
#     ggplot2::theme(legend.position="none") +
#     ggplot2::geom_vline(xintercept = QTL_Peak) +
#     ggplot2::facet_grid(geno_group ~ ., scales = "free", space = "free") + 
#     ggplot2::theme(panel.spacing = unit(1, "lines"))

print(p1)
# cowplot::plot_grid(p1, p2, ncol=2)

# ggplotly(p1) # cannot get the grid panel size to maintain the scale. to be fixed.
```
<br>
<br>

```{r, eval=split_by_div, results="asis"}
cat("\n## Phenotype x Genotype split by divergent region")
cat("  \n")
```

```{r, eval=split_by_div}
pxg_df <- processed_mapping %>%
    na.omit() %>% # probably the na.omit is not needed any more
    dplyr::filter(CHROM==QTL_Chrom, peakPOS==QTL_Peak) %>%  
    dplyr::mutate(value=as.numeric(value)) %>% 
    dplyr::inner_join(ref_alt_annotation, by="strain") %>% 
    dplyr::group_by(allele, is_div) %>%
    dplyr::mutate(mean_pheno = mean(value, na.rm = T))


pxg_plot <- pxg_df %>% 
    ggplot2::ggplot()+
    ggplot2::aes(x = factor(allele, levels = c(-1,1), labels = c("REF","ALT")))+
    ggbeeswarm::geom_beeswarm(cex=3, priority='density',
                  aes(y = value, plotly_label=strain),
                  shape = 21, 
                  fill = "gray50",
                  size = 2)+
    ggplot2::geom_point(aes(y = mean_pheno), 
               fill = "red", 
               size = 3, 
               shape = 25)+
    ggplot2::theme_bw(15)+
    ggplot2::labs(y = trait_name,
         x = "Genotype") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(size = 14),
          axis.text.y = ggplot2::element_text(size = 14),
          axis.title.x = ggplot2::element_text(size = 16, face = "bold", color = "black", vjust = -0.3),
          axis.title.y = ggplot2::element_text(size = 16, face = "bold", color = "black", vjust = -0.3),
          strip.text.x = element_text(size = 14, face = "bold"))+
    ggplot2::theme(legend.position = "none") +
    ggplot2::facet_wrap(~is_div, labeller = labeller(is_div = 
                                                c("-1" = "no div data",
                                                  "0" = "not divergent",
                                                  "1" = "divergent from N2")) )

plotly::ggplotly(pxg_plot, tooltip = c("plotly_label", "y"))
```



## Fine mapping plot
```{r, out.height=450, out.width=1000}
#################### Generate fine mapping plots from Finemap_QTL_Intervals.R

# load data. File name format is PC1.II.7319019.8506109_prLD_df.tsv
prLD_file_name <- paste0("Fine_Mappings/Data/",trait_name,".",QTL_Chrom,".",QTL_Region_start,".",QTL_Region_end,"_prLD_df.tsv")

pr_roi_ld <- read.delim(prLD_file_name, stringsAsFactors=FALSE) %>% 
    dplyr::mutate(marker = stringr::str_replace(marker, "_", ":")) 

peak_roi_marker <- pr_roi_ld %>%
    dplyr::filter(POS == QTL_Peak)

ld_plot <- pr_roi_ld %>% 
    ggplot2::ggplot(.) +
    ggplot2::aes(x = POS/1e6, plotly_label_ld = marker) +
    ggplot2::geom_point(aes(color = ld_r2, y = value), shape = 17, size = 2) +
    ggplot2::geom_point(data = peak_roi_marker, aes(y = value), shape = 23, size = 2, fill = "red") +
    ggplot2::scale_color_viridis_c(name = "LD R^2") +
    ggplot2::theme_bw(15)+
    ggplot2::labs(x = "Genomic Position (Mb)",
         y = "-log[10]p") 


# save plot to file
# ggsave(ld_plot, filename = paste0(QTL_peak, "_finemap_plot.pdf"),
#        height = 4,
#        width = 12)
# 

# interactive plots
plotly::ggplotly(ld_plot, tooltip = c("plotly_label_ld", "y")) # %>% toWebGL()

```



## Gene plot
```{r, out.height=450, out.width=1000}
######################## Generate gene plot from plot_genes.R
# create table to store gene related information
gene_info <- tidy_genes_in_region %>%
    dplyr::filter(QTL_INTERVAL_START == QTL_Region_start & !is.na(WBGeneID)) %>%
    dplyr::select(GENE_NAME, WBGeneID, CHROM, STRAND, TRANSCRIPTION_START_POS, TRANSCRIPTION_END_POS, QTL_INTERVAL_START, QTL_INTERVAL_END) %>% 
    dplyr::distinct()

# for each gene, keep the most significant overlapping variant log10p
# this is used as y-axix of gene position
gene_df <- tidy_genes_in_region %>%
    dplyr::filter(QTL_INTERVAL_START == QTL_Region_start & !is.na(WBGeneID)) %>%
    dplyr::distinct(MARKER, GENE_NAME, WBGeneID, PEAK_MARKER, CHROM, STRAND, TRANSCRIPTION_START_POS, TRANSCRIPTION_END_POS, 
                    QTL_INTERVAL_START, QTL_INTERVAL_END, VARIANT_LOG10p) %>%
    dplyr::group_by(WBGeneID) %>% 
    dplyr::summarise(VARIANT_LOG10p = max(VARIANT_LOG10p, na.rm = T)) %>% 
    dplyr::inner_join(gene_info) %>% 
    dplyr::mutate(STRAND = ifelse(STRAND=="+", "Gene_on_plus_strand", "Gene_on_minus_strand"))


peak_variant <- QTL_Peak

variant_df <- tidy_genes_in_region %>%
    dplyr::filter(QTL_INTERVAL_START == QTL_Region_start) %>%
    dplyr::distinct(CHROM, POS, VARIANT_LOG10p, VARIANT_IMPACT, REF, ALT) %>% 
    dplyr::mutate(MARKER_pos = paste0(CHROM, POS, " REF:ALT ", REF, ":", ALT))

variant_df$VARIANT_IMPACT[is.na(variant_df$VARIANT_IMPACT)] <- "Intergenic"

xs <- unique(gene_df$QTL_INTERVAL_START)
xe <- unique(gene_df$QTL_INTERVAL_END)
gc <- unique(gene_df$CHROM)

max_logp <- unique(max(variant_df$VARIANT_LOG10p))/150

gene_plot <- ggplot2::ggplot(gene_df) +
    ggplot2::geom_vline(aes(xintercept = peak_variant/1e6),
               linetype=3, color = "cyan")+
    ggplot2::geom_segment(aes(x = ifelse(STRAND == "Gene_on_plus_strand", TRANSCRIPTION_START_POS/1e6, TRANSCRIPTION_END_POS/1e6),
                     xend = ifelse(STRAND == "Gene_on_plus_strand", TRANSCRIPTION_END_POS/1e6, TRANSCRIPTION_START_POS/1e6),
                     y = VARIANT_LOG10p,
                     yend = VARIANT_LOG10p,
                     color = STRAND,
                     plotly_label_gene_name = GENE_NAME,
                     plotly_label_gene_ID = WBGeneID),
                 size = 1.5) +
    ggplot2::geom_segment(data = variant_df,
                 aes(x = POS/1e6,
                     xend = POS/1e6,
                     y = VARIANT_LOG10p+max_logp,
                     yend = VARIANT_LOG10p-max_logp,
                     color = VARIANT_IMPACT, 
                     plotly_label_variant = MARKER_pos)) +
    ggplot2::scale_color_manual(values = c("Gene_on_plus_strand" = "#3bb5c4",
                                  "Gene_on_minus_strand" = "#341d8d",
                                  "MODIFIER" = "gray50",
                                  "LOW" = "gray30",
                                  "MODERATE" = "orange",
                                  "HIGH" = "red",
                                  "Intergenic" = "gray80"),
                       name = "Variant Impact")+
    ggplot2::labs(x = "Genomic Position (Mb)",
         y = "-log[10]p") +
    ggplot2::theme_bw(18)+
    ggplot2::xlim(c(xs/1e6, xe/1e6)) +
    ggplot2::theme(legend.position = "top",
          panel.grid = element_blank())


# save plot to file
# ggsave(gene_plot,
#       filename = paste0(trait_name, ".", gc, ".",  xs, ".", xe, "_gene_plot.pdf"),
#        height=10, width = 14)


# interactive plot
plotly::ggplotly(gene_plot, tooltip = c("plotly_label_gene_name", "plotly_label_gene_ID", "plotly_label_variant")) %>% toWebGL()


# gene list for html output
# gene_info <- tidy_genes_in_region %>%
#     dplyr::filter(QTL_INTERVAL_START == QTL_Region_start & !is.na(WBGeneID)) %>%
#     dplyr::select(GENE_NAME, WBGeneID, CHROM, STRAND, TRANSCRIPTION_START_POS, TRANSCRIPTION_END_POS) %>% 
#     dplyr::distinct() %>% 
#     mutate(WBGeneID = paste0('<a href=https://wormbase.org/species/c_elegans/gene/', WBGeneID, '>', WBGeneID,'</a>' ) ) 
# 
# DT::datatable(gene_info, escape = FALSE)

```


<br>
<br>

#### All variants in interval

```{r}
# All variants in interval
var_info <- tidy_genes_in_region %>%
    dplyr::filter(QTL_INTERVAL_START == QTL_Region_start & !is.na(WBGeneID)) %>%
    dplyr::select(MARKER, REF, ALT,WBGeneID, VARIANT_IMPACT, VARIANT_LOG10p:STRAIN_GENOTYPE) %>%
    dplyr::distinct() %>%
    mutate(WBGeneID = paste0('<a href=https://wormbase.org/species/c_elegans/gene/', WBGeneID, '>', WBGeneID,'</a>' ) )

DT::datatable(var_info, escape = FALSE)



```

<br>
<br>

## Please kindly cite the following publications
* Cook, D.E., Zdraljevic, S., Roberts, J.P., and Andersen, E.C. (2016). CeNDR, the Caenorhabditis elegans natural diversity resource. Nucleic Acids Res.
* Zdraljevic, S., Fox, B.W., Strand, C., Panda, O., Tenjo, F.J., Brady, S.C., Crombie, T.A., Doench, J.G., Schroeder, F.C., and Andersen, E.C. (2019). Natural variation in C. elegans arsenic toxicity is explained by differences in branched chain amino acid metabolism. Elife 8.
* Lee, D., Zdraljevic, S., Stevens, L., Wang, Y., Tanny, R.E., Crombie, T.A., Cook, D.E., Webster, A.K., Chirakar, R., Ryan Baugh, L., et al. (2020). Balancing selection maintains ancient genetic diversity in C. elegans.
* haplotype paper

<br>
<br>

```{r}
sessionInfo()
```
